### PR TITLE
Initial Blue Green Strategy

### PR DESCRIPTION
**Pull request by dthomson25**

### CODE REVIEW

File: controller/controller.go [hunk 1]
```diff
@@ -154,16 +158,16 @@ func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
 	defer c.workqueue.ShutDown()
 
 	// Start the informer factories to begin populating the informer caches
-	klog.Info("Starting Foo controller")
+	klog.Info("Starting Rollout controller")
 
 	// Wait for the caches to be synced before starting workers
 	klog.Info("Waiting for informer caches to sync")
-	if ok := cache.WaitForCacheSync(stopCh, c.deploymentsSynced, c.foosSynced); !ok {
+	if ok := cache.WaitForCacheSync(stopCh, c.replicaSetSynced, c.rolloutsSynced); !ok {
```
Comment (alexmt) [inline controller/controller.go]: I've tried to create rollout without `strategy` and controller never starts. WaitForCacheSync stuck forever

```
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  name: example-rollout
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.15.4
        ports:
        - containerPort: 80
  minReadySeconds: 30
  revisionHistoryLimit: 3
```

File: utils/replicaset/replicaset.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	return allRSs
+}
+
+// NewRSNewReplicas calculates the number of replicas a Rollout's new RS should have.
+// When one of the followings is true, we're rolling out the deployment; otherwise, we're scaling it.
+// 1) The new RS is saturated: newRS's replicas == deployment's replicas
+// 2) Max number of pods allowed is reached: deployment's replicas + maxSurge == all RSs' replicas
+func NewRSNewReplicas(rollout *v1alpha1.Rollout, allRSs []*appsv1.ReplicaSet, newRS *appsv1.ReplicaSet) (int32, error) {
+	switch rollout.Spec.Strategy.Type {
+	case v1alpha1.BlueGreenRolloutStrategyType:
```
[the remark this answers is not in the document]
Comment (dthomson25) [reply utils/replicaset/replicaset.go]: My concern with blue green is that it requires an active service, and it won't be set if the Strategy field is empty.  What do you think about using a rolling-update once it's implemented as a default and have an invalid spec condition that says that the strategy needs to be filled out in the meanwhile.

**Comment by dthomson25** [comment]
I'm merging in this pr and I will address this feedback in future PRs

File: cmd/rollouts-controller/main.go [hunk 3]
```diff
@@ -56,22 +40,23 @@ func main() {
 		klog.Fatalf("Error building kubernetes clientset: %s", err.Error())
```
[the remark this answers is not in the document]
Comment (dthomson25) [reply cmd/rollouts-controller/main.go]: Addressed when we switched away from klog: https://github.com/argoproj/argo-rollouts/pull/7

**Pull request merged**